### PR TITLE
feat: encoder pro max

### PR DESCRIPTION
**Pull request by clearlysid**
This PR is a big one. It gets rid of FFmpeg in favour of OS-based encoders for the preview generation. This makes our binary even smaller, and makes the distribution and bundling proceses easier. Under the hood, It uses a custom Swift package for macOS and windows-capture encoder for Windows

TODOs:
1. clean up swift code
2. create BGRA encoding for mac encoder (currently it only works on YUV)
3. improve quality, colour config and codec of mac encoder
4. try building final binary

### CODE REVIEW

**Comment by clearlysid** [comment]
For the macOS encoder, we can use my https://github.com/clearlysid/bh-encode/tree/main as a reference point for testing. Currently the problem is that I have only created the encoder for YUVFrames captured in Mac. However, for Micro's usecase Gifski will prefer to have BGRA (i think, need to confirm).

The encoder will not work until we support that format also. This PR cannot be merged to main until that feature is complete.

**Comment by clearlysid** [comment]
oops

Screenshot 2024-04-23 at 12 17 03 AM

**Comment by clearlysid** [comment]
atleast it is fast 😂

**Comment by anubhavitis** [comment]
Not blurry at my end 🤔 
Screenshot 2024-04-23 at 12 33 44 AM

**Comment by anubhavitis** [comment]
Gif encoding is working as well :)
HM-240423-1234AM

File: src-tauri/build.rs [hunk 1]
```diff
@@ -1,30 +1,9 @@
...
-        println!("cargo:rustc-link-lib=framework=AudioToolbox");
-        println!("cargo:rustc-link-lib=framework=CoreAudio");
-        println!("cargo:rustc-link-lib=framework=Security");
-        println!("cargo:rustc-link-lib=framework=VideoToolbox");
-
-        link("z", ffmpeg_link_mode);
-        link("x264", ffmpeg_link_mode);
-        link("bz2", ffmpeg_link_mode);
-    }
+    swift_rs::SwiftLinker::new("11")
```
Comment (anubhavitis) [inline src-tauri/build.rs:4]: why "11" in const with understandable name.

File: src-tauri/src/recorder/encoder/mac/enc-swift/src/lib.swift [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+@_cdecl("encoder_init")
+func encoderInit(_ width: Int, _ height: Int, _ outFile: SRString) -> Encoder {
+    return Encoder(
+        width,
+        height,
+        URL(fileURLWithPath: outFile.toString())
+    )
+}
+
+func createCvPixelBufferFromYuvFrameData(
```
Comment (anubhavitis) [inline src-tauri/src/recorder/encoder/mac/enc-swift/src/lib.swift:62]: this is a util func for  encoderIngestYuvFrame
can be moved to separate file. Similarly util func for bgra as well.

File: src-tauri/build.rs (hunk 1 above)
Comment (clearlysid) [reply src-tauri/build.rs:4]: solved here: https://github.com/helmerapp/henx/commit/5ab39698957f7cfdc7098b6c05bdfc45d8ad31ba

File: src-tauri/src/recorder/encoder/mac/enc-swift/src/lib.swift (hunk 2 above)
Comment (clearlysid) [reply src-tauri/src/recorder/encoder/mac/enc-swift/src/lib.swift:62]: solved here: https://github.com/helmerapp/henx/commit/b6a774deca7a3d95f31d6192517805493bc3e188

File: src-tauri/src/main.rs [hunk 3]
```diff
@@ -84,17 +84,15 @@ fn main() {
...
 
-            // let first_run = true;
-            let first_run = store
-                .get("first_run".to_string())
-                .unwrap_or(&serde_json::Value::Bool(true))
-                .as_bool()
-                .unwrap();
-
-            let recording_permission: bool = scap::has_permission();
+            let first_run = true;
```
Comment (Pranav2612000) [inline src-tauri/src/main.rs:87]: Is this change expected?

**Comment by Pranav2612000** [comment]
> oops
> 
> Screenshot 2024-04-23 at 12 17 03 AM

Was this resolved? If not, I can help.

**Review by Pranav2612000 — APPROVED** [review]

File: src-tauri/src/main.rs (hunk 3 above)
Comment (clearlysid) [reply src-tauri/src/main.rs:87]: yes, the has_permission was preventing final (prod) builds from even opening properly on mac 😅
have removed it from the code, but on my list to investigate today: https://discord.com/channels/1044321624545046569/1194965312676827148/1232288042203611136

**Comment by clearlysid** [comment]
It is resolved on my machine, yup. Changed `byteStride`  to `width * 4` like we'd done somewhere in scap. The Stride somehow seems to be platform dependent though, I never figured out how it works.

For eg: in @anubhavitis's machine it was working properly even before I made the stride fix. Does that mean his version would be broken now? 🤔

**Comment by Pranav2612000** [comment]
> It is resolved on my machine, yup. Changed `byteStride` to `width * 4` like we'd done somewhere in scap. The Stride somehow seems to be platform dependent though, I never figured out how it works.
> 
> For eg: in @anubhavitis's machine it was working properly even before I made the stride fix. Does that mean his version would be broken now? 🤔

We can get on a call to debug this if this doesn't work.
I think the ` * 4 ` was for each of `RGBA`. 
If I'm not wrong the problem arises because the capturer adds extra black pixels if the width is not a multiple of 4. Trying to move the original width in this scenario would mess everything up.

Again, if it's working no need to touch it. But if we encounter errors, I'll be happy to help. I'll also test it out with different values.

**Pull request merged**